python/mypy_protobuf.py

```diff
@@ -643,34 +733,54 @@ def _add_messages(messages, prefix, _fd):
             _add_enums(fd.enum_type, start_prefix, fd)
 
 
-def main():
-    # type: () -> None
-    # Read request message from stdin
-    if six.PY3:
-        data = sys.stdin.buffer.read()
-    else:
-        data = sys.stdin.read()
+def wrap_code_generator(func):
+    # type: (Callable[[plugin_pb2.CodeGeneratorRequest, plugin_pb2.CodeGeneratorResponse], None]) -> Callable[[], None]
+    @wraps(func)
+    def wrapper():
+        # type: () -> None
+        # Read request message from stdin
+        if six.PY3:
+            data = sys.stdin.buffer.read()
+        else:
+            data = sys.stdin.read()
 
-    # Parse request
-    request = plugin_pb2.CodeGeneratorRequest()
-    request.ParseFromString(data)
+        # Parse request
+        request = plugin_pb2.CodeGeneratorRequest()
+        request.ParseFromString(data)
 
-    # Create response
-    response = plugin_pb2.CodeGeneratorResponse()
+        # Create response
+        response = plugin_pb2.CodeGeneratorResponse()
 
-    # Declare support for optional proto3 fields
-    response.supported_features |= (  # type: ignore[attr-defined]  # https://github.com/dropbox/mypy-protobuf/issues/158
-        plugin_pb2.CodeGeneratorResponse.FEATURE_PROTO3_OPTIONAL  # type: ignore[attr-defined]
-    )
+        # Declare support for optional proto3 fields
+        response.supported_features |= (  # type: ignore[attr-defined]  # https://github.com/dropbox/mypy-protobuf/issues/158
+            plugin_pb2.CodeGeneratorResponse.FEATURE_PROTO3_OPTIONAL  # type: ignore[attr-defined]
+        )
+
+        func(request, response)
+
+        # Serialise response message
+        output = response.SerializeToString()
+
+        # Write to stdout
+        if six.PY3:
+            sys.stdout.buffer.write(output)
+        else:
+            sys.stdout.write(output)
 
+    return wrapper
```

Comment:
Slight preference for using contextlib.contextmanager over the wrapped() - but not going to block the diff on it (I can fix it later). I always find it confusing to follow the control flow on wrapped functions like this.